Broker support

smurfix opened the pull request:
Here's a testcase for the built-in broker, along with assorted bugfixes to get it to work.

smurfix committed:
Support multiple-valued properties.

Required for subscription IDs in PUBLISH packets.

smurfix committed:
Speed up topic matching.

This special-cases
* a topic with no wildcard
* a topic with only a trailing '/#' (or just a '#')

Also, matching is sped up by checking for (and chopping off)
the pattern's static prefix.

smurfix committed:
Add a flag whether the broker supports subscription IDs.

smurfix committed:
Add subscription ID

to Subscription class and Subscribe message

smurfix committed:
Replace the client state machine's subscription handling.

The state machine no longer tries to remember which subscriptions
are used how often, thus no longer skips subscribing to topics it
already knows are subscribed.

Reason: The state associated with a topic includes the subscription
flags (retain, no_local), the subscription ID, and the QoS. The user
may want to update that state.

Remembering state is thus left to the user.

smurfix committed:
Use subscription IDs when available.

smurfix committed:
Add a test to check messages processing with overlapping subscriptions.

smurfix committed:
dup subscription processing dropped

smurfix committed:
Remove test_higher_than_max_qos

The check which this test tests for has been removed
because its logic was broken.

smurfix committed:
The broker doesn't do subscription IDs (for now)

smurfix committed:
Whitespace fixes as per review

Co-authored-by: Alex Grönholm <alex.gronholm@nextday.fi>

smurfix committed:
Docstring fix

smurfix committed:
Types: split Pattern to its own class

required for sane subscription handling in the broker

smurfix committed:
Don't call Stamina synchronously!

smurfix committed:
broker: return the port if it was zero

very useful for testing and dynamic brokers

smurfix committed:
async_client: need to ignore exceptions when sending, too

smurfix committed:
get the async broker to work

add/fix subscription handling

smurfix committed:
Don't play with ExitStack

Trio's dreaded long-winded "you mixed up your scopes" crash message
strikes again.

Just Say No to playing with ExitStack. It tends to not end well.

smurfix committed:
Merge branch 'bugs' into broker

smurfix committed:
Empty client IDs are illegal

smurfix committed:
always send a PUBACK

smurfix committed:
Oops

smurfix committed:
No that's not an error

smurfix committed:
Fix forwarding packets

We can't simply queue the packet to multiple places because
(a) the QoS might be different
(b) each client has its own packet ID space

smurfix committed:
Add a basic broker test

smurfix committed:
ruff format+check

plus some dead code

smurfix committed:
Fix the other tests

smurfix committed:
mypy 1

smurfix committed:
Merge branch 'subid' into broker

smurfix committed:
mypy checks (I hope)

smurfix committed:
client publish: pass user_properties

agronholm commented:
Maybe you should install `pre-commit` locally so you don't have to hope the GitHub CI passes?

smurfix committed:
Memoryviews are fickle

This raised a

BufferError: Existing exports of data: object cannot be re-sized

exception. I couldn't find the memoryview reference that causes this
(FWIW it sometimes happens when you transfer a user property)
so let's just use the alternate solution.

smurfix committed:
Test that user properties are passed

smurfix commented:
Huh. According to attrs' documentation this should be valid.

agronholm commented:
Are the argument types wrong? Does pre-commit pass for you locally?
